src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1529,6 +1529,89 @@ JvmtiEnvBase::is_in_thread_list(jint count, const jthread* list, oop jt_oop) {
   return false;
 }
 
+class VM_SetNotifyJvmtiEventsMode : public VM_Operation {
+private:
+  bool _enable;
+
+  // Iterates over all JavaThread's, counts VTMS transitions and restores
+  // jt->jvmti_thread_state() and jt->jvmti_vthread() for VTMS transition protocol.
+  int count_transitions_and_correct_jvmti_thread_states() {
+    int count = 0;
+
+    for (JavaThread* jt : ThreadsListHandle()) {
+      oop  ct_oop = jt->threadObj();
+      oop  vt_oop = jt->vthread();
+      JvmtiThreadState* jt_state = jt->jvmti_thread_state();
+      JvmtiThreadState* ct_state = java_lang_Thread::jvmti_thread_state(jt->threadObj());
+      JvmtiThreadState* vt_state = vt_oop != nullptr ? java_lang_Thread::jvmti_thread_state(vt_oop) : nullptr;
+      bool virt = vt_oop != nullptr && vt_oop != ct_oop;
+
+      if (jt->is_in_VTMS_transition()) {
+        count++;
+      }
+      // Correct jt->jvmti_thread_state() and jt->jvmti_vthread() if necessary.
+      // It was not maintained while notifyJvmti was disabled.
+      if (jt_state != ct_state && jt_state != vt_state) {
```

Comment:
Is it possible that jt_state == ct_state while the virtual thread is executed or vice versa? Just because jvmtt_state is outdated.
Shouldn't we always update (set to null) link/ jvmti_vthread if _enabled  == true?

Reply:
I've updated this part. Please. let me know if you still have some questions.